src/main/java/com/bobocode/hoverla/bring/context/ConfigBasedBeanDefinition.java

```diff
@@ -0,0 +1,172 @@
+     *                                             or bean method is not marked as {@link Bean}
+     * @throws NullPointerException                when any of arguments passed is null
+     */
+    public ConfigBasedBeanDefinition(Object configInstance, Method beanMethod) {
+        validate(configInstance, beanMethod);
+
+        log.debug("Creating {} from method '{}'", ConfigBasedBeanDefinition.class.getSimpleName(), beanMethod);
+        this.configInstance = configInstance;
+        this.beanMethod = beanMethod;
+
+        log.trace("Getting name for bean");
+        this.name = getName(beanMethod);
+        log.trace("The name is '{}'", name);
+
+        log.trace("Getting type for '{}' bean", name);
+        this.type = getType(beanMethod);
+        log.trace("The type is '{}'", type);
+
+        log.trace("Getting dependencies for '{}' bean", name);
+        this.dependencies = getDependencies(beanMethod);
+        log.trace("Dependencies for '{}' bean are {}", name, dependencies);
+    }
+
+    private static void validate(Object configInstance, Method method) {
+        log.debug("Validating arguments passed to create ConfigBased");
+        Objects.requireNonNull(configInstance, "Configuration class instance is null");
+        Objects.requireNonNull(method, "Configuration method to create bean");
+
+        checkClassMarkedAsConfig(configInstance);
+        checkMethodMarkedAsBean(method);
+    }
+
+    private static void checkClassMarkedAsConfig(Object configInstance) {
+        if (!configInstance.getClass().isAnnotationPresent(Configuration.class)) {
+            throw new BeanDefinitionConstructionException(
+                    "Configuration class instance passed is not marked as @%s".formatted(Configuration.class.getSimpleName())
+            );
+        }
+    }
+
+    private static void checkMethodMarkedAsBean(Method method) {
+        if (!method.isAnnotationPresent(Bean.class)) {
+            throw new BeanDefinitionConstructionException(
+                    "Configuration method to create bean is not marked as @%s".formatted(Bean.class.getSimpleName())
+            );
+        }
+    }
+
+    private static String getName(Method beanMethod) {
+        Bean annotation = beanMethod.getAnnotation(Bean.class);
+        String beanName = annotation.name();
+        if (beanName.isEmpty()) {
+            return beanMethod.getName();
+        }
+        return beanName;
+    }
+
+    private static Class<?> getType(Method beanMethod) {
+        return beanMethod.getReturnType();
+    }
+
+    private static Map<String, Class<?>> getDependencies(Method beanMethod) {
+        return Arrays.stream(beanMethod.getParameters())
+                     .collect(toMap(
+                             ConfigBasedBeanDefinition::getParameterName,
+                             Parameter::getType)
+                     );
+    }
+
+    private static String getParameterName(Parameter parameter) {
```

Comment:
How about this: 
``` 
return Optional.ofNullable(parameter.getAnnotation(Qualifier.class))
                .map(Qualifier::value)
                .orElse(parameter.getName());
```

Reply:
Yeah, it's a good option, but I don't think it's particularly better. Functional style is not a panacea here, I think.

Reply:
I prefer lambdas over multiple `if`, `else`, `return` etc. statements. But if you don't consider it better - you may leave it as is

Reply:
Yeah, me too, if it really makes code better.  love Optional API, but in this case it seems like just personal preference, taking into account that it's just a simple check and no multiple conditions are present.